shared/src/telemetry/bgworker.rs

```diff
@@ -0,0 +1,78 @@
+use pgrx::bgworkers::{BackgroundWorker, BackgroundWorkerBuilder, SignalWakeFlags};
+use pgrx::*;
+use std::ffi::CStr;
+use std::process;
+use std::thread;
+use std::time::{Duration, Instant};
+
+use crate::telemetry::posthog::read_and_send_telemetry_data;
+
+#[pg_guard]
+pub fn setup_telemetry_background_worker(extension_name: String) {
+    // A background worker to read and send telemetry data to PostHog.
+    BackgroundWorkerBuilder::new(&format!("{}_telemetry_worker", &extension_name))
+        // Must be the name of a function in this file.
+        .set_function("telemetry_worker")
+        // Must be the name of the extension it will be loaded from.
+        .set_library(&extension_name)
+        // We pass the extension name to retrieve the associated data directory to read telemetry data from.
+        .set_argument(extension_name.into_datum())
+        // Necessary for using plog!.
+        // Also, it doesn't seem like bgworkers will start without this.
+        .enable_spi_access()
+        // RecoveryFinished is the last available stage for bgworker startup.
+        // We wait until as late as possible so that we can make sure the
+        // paradedb.logs table is created, for the sake of using plog!.
+        .set_start_time(bgworkers::BgWorkerStartTime::RecoveryFinished)
+        .load();
+}
+
+#[pg_guard]
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn telemetry_worker(arg: pg_sys::Datum) {
+    // Convert Datum to CString
+    let text_ptr =
+        pg_sys::pg_detoast_datum(arg.cast_mut_ptr::<pg_sys::varlena>()) as *mut pg_sys::text;
+    let c_str = CStr::from_ptr(pg_sys::text_to_cstring(text_ptr));
+
+    // Convert CStr to Rust String
+    let rust_string = match c_str.to_str() {
```

Comment:
You can avoid doing all this error unwrapping by moving this function to a Rust function, using `thiserror`, and then calling that function with a single `.unwrap_or_else()`. An example of this is in pg_analytics `tableam/insert.rs`.